whisperv6/whisper.go

```diff
@@ -407,12 +407,40 @@ func (whisper *Whisper) AllowP2PMessagesFromPeer(peerID []byte) error {
 // which are not supposed to be forwarded any further.
 // The whisper protocol is agnostic of the format and contents of envelope.
 func (whisper *Whisper) RequestHistoricMessages(peerID []byte, envelope *Envelope) error {
+	return whisper.RequestHistoricMessagesWithTTL(peerID, envelope, 0)
+}
+
+func (whisper *Whisper) RequestHistoricMessagesWithTTL(peerID []byte, envelope *Envelope, ttl time.Duration) error {
 	p, err := whisper.getPeer(peerID)
 	if err != nil {
 		return err
 	}
+	whisper.envelopeFeed.Send(EnvelopeEvent{
+		Peer:  p.peer.ID(),
+		Hash:  envelope.Hash(),
+		Event: EventMailServerRequestSent,
+	})
 	p.trusted = true
-	return p2p.Send(p.ws, p2pRequestCode, envelope)
+	err = p2p.Send(p.ws, p2pRequestCode, envelope)
+	if ttl != 0 {
+		go whisper.expireRequestHistoricMessages(p.peer.ID(), envelope.Hash(), ttl)
+	}
+	return err
+}
+
+func (whisper *Whisper) expireRequestHistoricMessages(peer enode.ID, hash common.Hash, ttl time.Duration) {
```

Comment:
I don't quite understand this method. It always sends `EventMailServerRequestExpired` event unless `whisper` is stopped.

Also, I have been looking where these events are used and I guess we have an embarrassing bug: https://github.com/status-im/whisper/blob/14e1bbfd9ba956e7fc7649e815aa19e4386b26da/whisperv6/mailserver_response.go#L79

Reply:
yes, it always sends expire event. this event can be catched by listener. so for example if you catch it before completed event, then request is clearly expired. we do the same for regular envelopes

Reply:
what is the bug? i am not sure that i understood

Reply:
> yes, it always sends expire event. this event can be catched by listener. so for example if you catch it before completed event, then request is clearly expired. we do the same for regular envelopes

We need to better describe it here because it looks suspicious and this knowledge is not obvious.

> what is the bug? i am not sure that i understood

The method is called `tryCreateMailServerRequestFailedEvent` but it creates `EventMailServerRequestCompleted` event.

Reply:
> We need to better describe it here because it looks suspicious and this knowledge is not obvious.

sure, added a comment to an event description